### ISSUE TITLE
False-positive error when using `Callable` in a PEP 604 type alias in a stub file with `--python-version=3.9`

### ISSUE THREAD

**Issue by AlexWaygood** [python/mypy#14158]
**Bug Report**

In a stub file, with `--python-version=3.9`, mypy emits false positive errors for the following type aliases:

```python
from typing import Callable
from typing_extensions import TypeAlias

X = int | Callable[[], str | bool]
XX: TypeAlias = int | Callable[[], str | bool]
Y = int | Callable[[str | bool], str]
YY: TypeAlias = int | Callable[[str | bool], str]
```

For all of these, mypy's error message is:

```
error: Unsupported left operand type for | ("Type[str]")  [operator]
```

Note that this snippet works fine in a `.py` or `.pyi` file with `--python-version 3.10`. It's only `--python-version 3.9` with a `.pyi` file that has the bug.

**To Reproduce**

1. `cd` into an up-to-date local clone of mypy with an editable install.
2. Copy the above snippet into a `test.pyi` file.
3. Run `mypy test.pyi --python-version 3.9`

**Expected Behavior**

No error should be emitted. PEP 604 syntax should be legal in stub files, even with `--python-version 3.9`.

**Actual Behavior**

False-positive errors are emitted.

**Your Environment**

- Mypy version used: https://github.com/python/mypy/commit/c660354846688ff8158d0f0178eb298171b74f5b
- Mypy command-line flags: `--python-version 3.9`
- Python version used: 3.10

Cc. @ilevkivskyi (x-ref #12393).

**Comment by ilevkivskyi** [comment]
cc @JukkaL who recently worked on this.

**Comment by adam-grant-hendry** [comment]
Question: I have this issue in Python 3.8 in a normal `.py` file, but am able to get around it by 'stringifying' the type:

```python
from __future__ import annotations

from typing import TYPE_CHECKING, Callable

from poetry.console.application import Application

if TYPE_CHECKING:
    import sys

    from poetry.poetry import Poetry

    if sys.version_info >= (3, 10):
        from typing import TypeAlias
    else:
        from typing_extensions import TypeAlias

ApplicationFactory: TypeAlias = 'Callable[[Poetry | None], Application]'
PoetryFactory: TypeAlias = 'Callable[[str | None], Poetry]'
```

Is this intended behavior for 3.8 with `from __future__ import annotations`, or should I not need to stringify the type?

**Comment by AlexWaygood** [comment]
@adam-grant-hendry that's correct behaviour from mypy. It'll fail at runtime on Python <3.10 if you don't stringify the alias, even with `from __future__ import annotations`.

`from __future__ import annotations` only impacts the way Python evaluates annotations at runtime; it doesn't have any impact on how Python evaluates ordinary assignments at runtime. Both  `ApplicationFactory` and `PoetryFactory` are using PEP-604 syntax in assignments rather than in annotations.

**Comment by adam-grant-hendry** [comment]
> that's correct behaviour from mypy

Excellent, thank you for confirming! Glad to know I did this properly 😃
